File: opencv/python/BallTracker/BallTracker.py
import numpy as np
import cv2
import os
from functools import partial
import collections
import copy

DEBUG = True
DEBUG_VERBOSE = True


"""
Static Video capture type. Can be a camera, a video, or an image.
"""
class CaptureType:
	Camera, Video, Image = range(3)

	@staticmethod
	def toString(type):
		if type == CaptureType.Camera:
			return 'Camera'
		elif type == CaptureType.Video:
			return 'Video'
		elif type == CaptureType.Image:
			return 'Image'
		else:
			return 'Unknown'


"""
Tracking parameters for BallTracker.
"""
class TrackingParameters:
	def __init__(self, hueLower=27, hueUpper=42, satLower=58, satUpper=256,
		valLower=44, valUpper=256, filterIterations=3, filterSigma=2,
		houghDp=1, houghThreshUpper=114, houghThreshLower=16, 
		houghMinRadius=3, houghMaxRadius=140):
		# Parameters
		self.HueLower = hueLower
		self.HueUpper = hueUpper
		self.SatLower = satLower
		self.SatUpper = satUpper
		self.ValLower = valLower
		self.ValUpper = valUpper
		self.FilterIterations = filterIterations
		self.FilterSigma = filterSigma
		self.HoughDP = houghDp
		self.HoughThreshUpper = houghThreshUpper
		self.HoughThreshLower = houghThreshLower
		self.HoughMinRadius = houghMinRadius
		self.HoughMaxRadius = houghMaxRadius
		# Other attributes
		self.ParameterChanged = False

	"""
	Event called when parameter is changed.
	"""
	def changeParameter(self,val, parameter):
		setattr(self,parameter,val)
		self.ParameterChanged = True
		if (DEBUG and DEBUG_VERBOSE):
			print parameter, ': ', val


	"""
	Prints all parameters.
	"""
	def printParameters(self):
		print "Tracking parameters:"
		paramDict = vars(self)
		oParamDict = collections.OrderedDict(sorted(paramDict.items()))
		houghParamDict = collections.OrderedDict()

		#print paramDict
		# Print 
		for k, v in oParamDict.items():
			if k.startswith('Hough'):
				houghParamDict[k] = v
			else:
				print "\t{}: {}".format(k,v)
		# Hough params
		for k, v in houghParamDict.items():
			print "\t{}: {}".format(k,v)

"""
Opens a camera, video, or image file and tracks the ball.
"""
class BallTracker:
	"""
	Constructor initializes BallTracker, parameters, and opens the file/device.
	"""
	def __init__(self, deviceOrFile=0, trackBall=True, logFile=None, camWidth=640, camHeight=480, pauseVideoFrames=False):
		self.TrackingEnabled = trackBall
		self.IsRunning = False
		self.InterfaceInitialized = False
		self.VideoWaits = pauseVideoFrames
		# Check whether device or file was given
		if (type(deviceOrFile) is str):
			if (deviceOrFile.lower().endswith(('.png', '.jpg', '.jpeg', '.bmp', '.gif',
				'.dib','.jp2', '.pbm', '.pgm', '.ppm', '.tiff', '.tif'))):
				self.CaptureType = CaptureType.Image
			elif (deviceOrFile.lower().endswith(('.avi','.mov','.mpeg'))):
				self.CaptureType = CaptureType.Video
			else:
				raise Exception('Unknown file extension for file: {}'.format(deviceOrFile))
			self.Input = os.path.abspath(deviceOrFile)
			if (not os.path.isfile(self.Input)):
				raise Exception('Could not find input file: {}'.format(self.Input))
		elif (type(deviceOrFile) in (int, float)):
			self.Input = deviceOrFile
			self.CaptureType = CaptureType.Camera
		else:
			raise 'Unknown input deviceOrFile given'

		# Open input
		self.Cap = cv2.VideoCapture(self.Input)
		if (not self.Cap.isOpened()):
			raise Exception('Could not open capture ({}): {}'.format(
				CaptureType.toString(self.CaptureType), self.Input))
		# Get/set capture information
		if (self.usingCamera()):
			# Set width and height for camera
			self.Cap.set(cv2.CAP_PROP_FRAME_WIDTH,camWidth);
			self.Cap.set(cv2.CAP_PROP_FRAME_HEIGHT,camHeight);
			# Printing info for debug mode
			if (DEBUG):
				print 'Camera: device ', self.Input
				print('Video properties:\n\tWidth: {}px\n\tHeight: {}px'.format(camWidth, camHeight))
				print '\tMax Width: {} px\n\tMax Height: {}px'.format(
					int(self.Cap.get(cv2.CAP_PROP_FRAME_WIDTH)),
					int(self.Cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))
				print '\tMode: ', self.Cap.get(cv2.CAP_PROP_MODE)
				print '\tBrightness: ', self.Cap.get(cv2.CAP_PROP_BRIGHTNESS)
				print '\tContrast: ', self.Cap.get(cv2.CAP_PROP_CONTRAST)
				print '\tSatur: ', self.Cap.get(cv2.CAP_PROP_SATURATION)
				print '\tHue: ', self.Cap.get(cv2.CAP_PROP_HUE)
				print '\tGain: ', self.Cap.get(cv2.CAP_PROP_HUE)
				print '\tExposure: ', self.Cap.get(cv2.CAP_PROP_EXPOSURE)
		else:
			# Find width and height for video/image
			typeName = CaptureType.toString(self.CaptureType)
			self.Width = int(self.Cap.get(cv2.CAP_PROP_FRAME_WIDTH))
			self.Height = int(self.Cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
			# Get FPS and frames for video
			if (self.usingVideo()):
				self.FPS = self.Cap.get(cv2.CAP_PROP_FPS)
				self.FrameCount = int(self.Cap.get(cv2.CAP_PROP_FRAME_COUNT))
				self.Length = self.FrameCount / self.FPS
			# Printing debug info
			if (DEBUG):
				print '{} file: {}', typeName, self.Input
				print('{} properties:\n\tWidth: {}px\n\tHeight: {}px'.format(
					typeName, self.Width, self. Height))
				if (self.usingVideo()):
					print '\tFPS: ', self.FPS
					print '\tFrames: ', self.FrameCount
					print '\tLength: ', self.Length

		# Create parameter class
		self.TrackingParameters = TrackingParameters()
		if (DEBUG):
			self.TrackingParameters.printParameters()
			print "Ball tracker initialized."
	# end of def __init__()

	"""
	Returns True if the BallTracker is using a video file.
	"""
	def usingVideo(self):
		return self.CaptureType == CaptureType.Video

	"""
	Returns True if the BallTracker is using an image file.
	"""
	def usingImage(self):
		return self.CaptureType == CaptureType.Image

	"""
	Returns True if the BallTracker is using a camera.
	"""
	def usingCamera(self):
		return self.CaptureType == CaptureType.Camera


	"""
	Initializes BallTracker interface with video windows and sliders for setting parameters.
	"""
	def initializeInterface(self):
		if (self.InterfaceInitialized):
			return

		# Create windows
		cv2.namedWindow('Camera', cv2.WINDOW_AUTOSIZE)
		cv2.namedWindow('Thresholded', cv2.WINDOW_AUTOSIZE)

		# Thresholded trackbars
		cv2.createTrackbar('HueLower','Thresholded',self.TrackingParameters.HueLower,
			255,partial(self.TrackingParameters.changeParameter, parameter='HueLower'))
		cv2.createTrackbar('HueUpper','Thresholded',self.TrackingParameters.HueUpper,
			255,partial(self.TrackingParameters.changeParameter, parameter='HueUpper'))
		cv2.createTrackbar('SatLower','Thresholded',self.TrackingParameters.SatLower,
			255,partial(self.TrackingParameters.changeParameter, parameter='SatLower'))
		cv2.createTrackbar('ValLower','Thresholded',self.TrackingParameters.ValLower,
			255,partial(self.TrackingParameters.changeParameter, parameter='ValLower'))
		cv2.createTrackbar('FilterIterations','Thresholded',self.TrackingParameters.FilterIterations,
			32,partial(self.TrackingParameters.changeParameter, parameter='FilterIterations'))
		cv2.createTrackbar('FilterSigma','Thresholded',self.TrackingParameters.FilterSigma,
			32,partial(self.TrackingParameters.changeParameter, parameter='FilterSigma'))

		# Hough circle trackbars
		minRadiusSliderMax = min(self.Height,self.Width)
		maxRadiusSliderMax = max(self.Height,self.Width)
		cv2.createTrackbar('HoughDP','Camera',self.TrackingParameters.HoughDP,
			16,partial(self.TrackingParameters.changeParameter, parameter='HoughDP'))
		cv2.createTrackbar('HoughThreshUpper','Camera',self.TrackingParameters.HoughThreshUpper,
			500,partial(self.TrackingParameters.changeParameter, parameter='HoughThreshUpper'))
		cv2.createTrackbar('HoughThreshLower','Camera',self.TrackingParameters.HoughThreshLower,
			100,partial(self.TrackingParameters.changeParameter, parameter='HoughThreshLower'))
		cv2.createTrackbar('HoughMinRadius','Camera',self.TrackingParameters.HoughMinRadius,
			minRadiusSliderMax,partial(self.TrackingParameters.changeParameter, parameter='HoughMinRadius'))
		cv2.createTrackbar('HoughMaxRadius','Camera',self.TrackingParameters.HoughMaxRadius,
			maxRadiusSliderMax,partial(self.TrackingParameters.changeParameter, parameter='HoughMaxRadius'))
		self.InterfaceInitialized = True
	# end of def initializeInterface

	"""
	Runs the BallTracker.
	"""
	def run(self):
		if (self.Cap is None or not self.Cap.isOpened()):
			return
		if (not self.InterfaceInitialized):
			self.initializeInterface()
		self.IsRunning = True
		frame_i = 0
		wantNextFrame = True # only used if usingVideo() and VideoWaits
		heldFrame = False
		# Read frame only once for an image
		if (self.usingImage()):
			ret, frame = self.Cap.read()
			origFrame = frame
			self.Cap.release()

		# Main Loop
		while(self.IsRunning):
			if (not self.usingImage()):
				if (self.usingVideo() and self.VideoWaits and not wantNextFrame):
					heldFrame = True
				else:
					ret, frame = self.Cap.read()
					wantNextFrame = False
					heldFrame = False
			# Loop video captures if we're at the last frame
			if (self.usingVideo()):
				if (self.VideoWaits and heldFrame):
					pass
				else:
					frame_i += 1
					if frame_i == self.Cap.get(cv2.CAP_PROP_FRAME_COUNT):
						frame_i = 0
						self.Cap.release()
						self.Cap = cv2.VideoCapture(self.Input)

			threshFrame = self.thresholdFrame(frame)
			found, (x, y), radius = self.detectBall(threshFrame)

			# Draw the ball if it was found
			newFrame = copy.deepcopy(frame)
			if (found):
				self.drawBall(newFrame, (x,y), radius)
				if (DEBUG and DEBUG_VERBOSE and not heldFrame):
					print "Found ball at ({}, {}), with radius {}".format(x,y, radius)
			else:
				if (DEBUG and DEBUG_VERBOSE and not heldFrame):
					ending = '.'
					if (self.usingVideo()):
						ending = ' at frame {} of {}'.format(frame_i, self.FrameCount)
					print "Did not find ball", ending

			# Show the videos
			cv2.imshow('Camera',newFrame)
			cv2.imshow('Thresholded', threshFrame)

			# Make sure capture is still open
			if (not self.usingImage() and not self.Cap.isOpened()):
				break

			# Check for a key press
			key = cv2.waitKey(1) & 0xFF

			if (key == ord('q')):
				break
			if (key == ord('n')):
				wantNextFrame = True
			if (key == ord('p')):
				self.TrackingParameters.printParameters()
		# end of while(self.IsRunning)

		self.IsRunning = False

		if (self.Cap and self.Cap.isOpened()):
			self.Cap.release()
		cv2.destroyAllWindows()
	# end of def run()

	"""
	Draws the ball in the frame at position (x,y) with given radius.
	"""
	def drawBall(self, frame, (x,y), radius, innerRadius=3, innerCircleColor=(0,255,0),
		outerCircleColor=(0,0,255)):
		# Draw inner circle (small dot at center of ball)
		cv2.circle(frame, (x,y),
			radius=innerRadius,
			color=innerCircleColor,
			thickness=-1,
			lineType=cv2.LINE_8,
			shift=0)
		# Draw outer circle
		cv2.circle(frame, (x,y),
			radius=radius,
			color=outerCircleColor,
			thickness=1,
			lineType=cv2.LINE_8,
			shift=0)

	# end of def drawBall()

	"""
	Thresholds each frame and returns the thresholded frame.
	"""
	def thresholdFrame(self, frame):
		# Convert color space 
		hsvFrame = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
		hsvRangeLower = (self.TrackingParameters.HueLower, self.TrackingParameters.SatLower,
			self.TrackingParameters.ValLower, 0)
		hsvRangeUpper = (self.TrackingParameters.HueUpper, self.TrackingParameters.SatUpper,
			self.TrackingParameters.ValUpper, 0)

		threshFrame = cv2.inRange(hsvFrame, lowerb=hsvRangeLower, upperb=hsvRangeUpper)
		# if (DEBUG and np.count_nonzero(threshFrame) > 0):
		# 	#print np.nonzero(threshFrame)
		# 	print "Found a non-zero threshFrame at: {}".format(cap.get(cv2.CAP_PROP_POS_MSEC))

		# Filtering
		threshFrame = cv2.dilate(threshFrame, kernel=None,
			iterations=self.TrackingParameters.FilterIterations)
		threshFrame = cv2.GaussianBlur(threshFrame, ksize=(0, 0),
			sigmaX=max(self.TrackingParameters.FilterSigma,1),
			sigmaY=max(self.TrackingParameters.FilterSigma,1))
		threshFrame = cv2.erode(threshFrame, kernel=None,
			iterations=self.TrackingParameters.FilterIterations)

		return threshFrame

	# end of def thresholdFrame()

	"""
	Processes the given thresholded frame and detects the ball.
	Returns: found, (x,y), radius
	"""
	def detectBall(self, threshFrame):
		# Find circles
		circles = cv2.HoughCircles(threshFrame,
			method=cv2.HOUGH_GRADIENT,
			dp=self.TrackingParameters.HoughDP,
			minDist=50,
			param1=max(self.TrackingParameters.HoughThreshUpper,1),
			param2=max(self.TrackingParameters.HoughThreshLower,1),
			minRadius=self.TrackingParameters.HoughMinRadius,
			maxRadius=self.TrackingParameters.HoughMaxRadius)

		# Detect ball by finding largest circle
		maxRadius = 0
		(x, y) = (0, 0)
		found = False
		if circles is not None:
			circles = circles[0,:,:] # strip off 3rd dim
			total = circles.shape[0] # count circles found
			#if (DEBUG):
			#	print "Found {} Hough circles".format(total)
			# Iterate over each and find the largest circle
			for i in range(0, total):
				c = np.transpose(circles[i,:])
				if (c[2] > maxRadius):
					maxRadius = c[2]
					x = c[0]
					y = c[1]
					found = True

		return found, (x,y), maxRadius

	# end of def detectBall()

# end of class BallTracker()

## *** Entry Point ***
if __name__ == "__main__":
    from optparse import OptionParser

    #parser = OptionParser("%prog [options] <inputVideoFileOrDevice>")
    #parser.add_option("-o", "--output", dest="output", default=None, help="output video")
    #(opts, args) = parser.parse_args()
    args = ['../../../video/tennis_ball2-cut.mov']
    if len(args) < 1:
    	parser.error("You must give an input video file.")

    # Create tracker and run it
    tracker = BallTracker(deviceOrFile=args[0], pauseVideoFrames=True)
    tracker.run()

File: opencv/cpp/BallTracker/BallTracker/BallTracker.cpp
#define _USE_MATH_DEFINES // to get M_PI
#include <iostream>
#include <stdexcept>
#include <stdint.h>
#include <cmath>
#include <functional>
#include "BallTracker.hpp"
#include "Utility.hpp"

// Intialize CaptureType struct names
const char* CaptureType::names[] = {"camera", "video", "image"};

// List of extensions for image files
const std::string imageFileExtensions[] = {".png", ".jpg", ".jpeg", ".bmp", ".gif", ".dib",
    ".jp2", ".pbm", ".pgm", ".ppm", ".tiff", ".tif"};
std::vector<std::string> imageFileExtensionList(imageFileExtensions, end(imageFileExtensions));

//-------------------------------- Member functions -----------------------------
BallTracker::BallTracker(int deviceNum, bool debug, bool debugVerbose, bool trackBall, int camWidth,
    int camHeight, TrackingParameters *trackingParameters) {
	isRunning = false;
	interfaceIsInitialized = false;
	videoIsPaused = false;
	foundBall = false;
	ballCenter = cv::Point2i(0,0);
	ballRadius = 0;
    this->deviceNum = deviceNum;
    trackingEnabled = trackBall;
    captureProperties["width"] = (float)camWidth;
    captureProperties["height"] = (float)camHeight;
    this->trackingParameters = trackingParameters;
    this->trackingParameters->debug = debugVerbose;
    captureType = CaptureType::camera;
    this->debug = debug;
    this->debugVerbose = debugVerbose;

    // Open the capture
    cap = cv::VideoCapture(this->deviceNum);
    if (!cap.isOpened())
        throw std::runtime_error("failed to capture from device");

    // Set capture properties
    cap.set(cv::CAP_PROP_FRAME_WIDTH, camWidth);
    cap.set(cv::CAP_PROP_FRAME_HEIGHT, camHeight);

    // Debug information
    if (debug) {
        std::cout << "Camera: device " << this->deviceNum << std::endl;
        std::cout << "Properties:" << std::endl << "\tWidth: " << captureProperties["width"] << std::endl
			<< "\tHeight: "<< captureProperties["height"] << std::endl;
        std::cout << "\tMax Width: " << cap.get(cv::CAP_PROP_FRAME_WIDTH) << std::endl;
        std::cout << "\tMax Height: " << cap.get(cv::CAP_PROP_FRAME_HEIGHT) << std::endl;
        /*
		std::cout << "\tMode: " << cap.get(cv::CAP_PROP_MODE) << std::endl;
        std::cout << "\tBrightness: " << cap.get(cv::CAP_PROP_BRIGHTNESS) << std::endl;
        std::cout << "\tContrast: " << cap.get(cv::CAP_PROP_CONTRAST) << std::endl;
        std::cout << "\tSaturation: " << cap.get(cv::CAP_PROP_SATURATION) << std::endl;
        std::cout << "\tHue: " << cap.get(cv::CAP_PROP_HUE) << std::endl;
        std::cout << "\tGain: " << cap.get(cv::CAP_PROP_GAIN) << std::endl;
        std::cout << "\tExposure: " << cap.get(cv::CAP_PROP_EXPOSURE) << std::endl;
		*/

        std::cout << "Ball tracker initialized." << std::endl;
    }
}

BallTracker::BallTracker(std::string fileName, bool debug, bool debugVerbose, bool trackBall,
    bool startVideoPaused, TrackingParameters *trackingParameters) {
	isRunning = false;
	frameNumber = 1;
	interfaceIsInitialized = false;
	videoIsPaused = false;
	foundBall = false;
    this->fileName = fileName;
    trackingEnabled = trackBall;
    videoIsPaused = startVideoPaused;
    this->trackingParameters = trackingParameters;
    this->trackingParameters->debug = debugVerbose;
    captureType = CaptureType::video;
    this->debug = debug;
    this->debugVerbose = debugVerbose;

    // Check if we have an image file using the file extension
    std::string fileNameExt = fileName.substr(fileName.find_last_of(".") + 1);
    if (std::find(imageFileExtensionList.begin(), imageFileExtensionList.end(),
        fileNameExt) != imageFileExtensionList.end()) {
        captureType = CaptureType::image;
    }

    // Open the capture
    cap = cv::VideoCapture(this->fileName);
    if (!cap.isOpened())
		throw std::runtime_error("Failed to capture from file");

    // Get capture properties
    captureProperties["width"] = (float)cap.get(cv::CAP_PROP_FRAME_WIDTH);
    captureProperties["height"] = (float)cap.get(cv::CAP_PROP_FRAME_HEIGHT);
    if (isUsingVideo()) {
        captureProperties["fps"] = (float)cap.get(cv::CAP_PROP_FPS);
        captureProperties["frameCount"] = (float)cap.get(cv::CAP_PROP_FRAME_COUNT);
        captureProperties["length"] = captureProperties["frameCount"] / captureProperties["fps"];
    }

    // Debug information
    if (debug) {
        std::cout << CaptureType::names[captureType] << " file: " << this->fileName << std::endl;
        std::cout << "Properties:" << std::endl
			<< "\tWidth: " << captureProperties["width"] << std::endl
			<< "\tHeight: " << captureProperties["height"] << std::endl;
        if (isUsingVideo()) {
            std::cout << "\tFPS: " << captureProperties["fps"] << std::endl;
            std::cout << "\tFrames: " << captureProperties["frameCount"] << std::endl;
            std::cout << "\tLength: " << captureProperties["length"] << std::endl;
        }
        std::cout << "Ball tracker initialized." << std::endl;
    }
}

BallTracker::~BallTracker() {
    if (cap.isOpened())
        cap.release();

    cv::destroyAllWindows();
}

bool BallTracker::isUsingVideo() {
    return captureType == CaptureType::video;
}

bool BallTracker::isUsingImage() {
    return captureType == CaptureType::image;
}

bool BallTracker::isUsingCamera() {
    return captureType == CaptureType::camera;
}

int BallTracker::getWidth() {
	return (int)captureProperties["width"];
}

int BallTracker::getHeight() {
	return (int)captureProperties["height"];
}


void BallTracker::run() {
	if (! cap.isOpened())
		return;

	if (! interfaceIsInitialized)
		initializeInterface();

	bool isRunning = true;
	bool wantNextFrame = true; // want a new frame
	bool haveNewFrame = false; // new frame read
	int result;
	cv::Mat frame(getHeight(),getWidth(), CV_8UC3), 
			frameFiltered(getHeight(),getWidth(), CV_8UC3), 
			frameBall(getHeight(), getWidth(), CV_8UC3);

	// Only load images once
	if(isUsingImage()) {
		result = cap.read(frame);
		haveNewFrame = true;
		cap.release();
	}
	
	backgroundSubtractor = cv::createBackgroundSubtractorMOG2();

	// Main loop
	while (isRunning) {
		// Read the frame if not using an single image
		if (!isUsingImage()) {
			if (videoIsPaused && !wantNextFrame) {
				// do nothing
			}
			else {
				result = cap.read(frame);
				wantNextFrame = false;
				haveNewFrame = true;
			}
		}
		frame.copyTo(frameBall);

		// Process frame by filtering and then detecting the ball.
		filterFrame(frame, frameFiltered);
		if (trackingEnabled) {
			foundBall = detectBall(frameFiltered, ballCenter, ballRadius);

			if (foundBall) {
				drawBall(frameBall,ballCenter,ballRadius);
			}
		}

		// Frame info shows in bottom corner of video.
		if (isUsingVideo()) {
			double textScale = 0.8f;
			int textThickness = 1;
			cv::Scalar textColor(255,160,160);
			char *pausedStr = "";
			if (videoIsPaused)
				pausedStr = " (paused)";
			sprintf(tempBuf, "Frame %u of %u%s.", frameNumber, (int)captureProperties["frameCount"], pausedStr);
			cv::putText(frameBall, std::string(tempBuf),cv::Point(0,(int)captureProperties["height"]-10),
				cv::FONT_HERSHEY_PLAIN, textScale, textColor, textThickness);
		}

		// Debug info prints into console
		if (debug && (trackingParameters->wasChanged() || haveNewFrame)) {
			tempBuf[0] = '.'; tempBuf[1] = '\0';

			if (isUsingVideo())
				sprintf(tempBuf, " at frame %u of %u.", frameNumber, (unsigned int)captureProperties["frameCount"]);

			if (foundBall)
				std::cout << "Found ball at " << ballCenter.x << ", " << ballCenter.y << " with radius "
					<< ballRadius << tempBuf << std::endl;
			else
				std::cout << "Did not find ball" << tempBuf << std::endl;
		}

		// Show the frames
		cv::imshow("Camera", frameBall);
		cv::imshow("Filtered", frameFiltered);

		// Get key presses
		int key = cv::waitKey(1) & 0xFF;
		if (key == 'q') // quits
			break;
		else if (key == 'n') // advances to next frame when paused
			wantNextFrame = true;
		else if (key == 'p')
			trackingParameters->printParameters(); // pause
		else if (key == 13) // enter - pauses a video or camera
			videoIsPaused = videoIsPaused ^ 1;

		// Advancing to next frame
		if (isUsingVideo() && !videoIsPaused)
			wantNextFrame = true;
		if (!isUsingImage() && wantNextFrame)
			frameNumber++;

		// Makes a video loop by releasing and opening a new capture.
		if (isUsingVideo() && frameNumber >= (int)captureProperties["frameCount"]) {
			frameNumber = 1;
			cap.release();
			cap = cv::VideoCapture(fileName);
		}
		haveNewFrame = false;
	} // while (isRunning())

	if (debug) {
		trackingParameters->printParameters();
		std::cout << "Ball tracker finished." << std::endl;
	}
}

void BallTracker::initializeInterface() {
	if (interfaceIsInitialized)
		return;

	// Create windows
    cv::namedWindow("Camera", cv::WINDOW_AUTOSIZE);
    cv::namedWindow("Filtered", cv::WINDOW_AUTOSIZE);
    cv::namedWindow("Controls", cv::WINDOW_NORMAL);

	// Filter window trackbars
	createTrackbar("hueLower", "Controls", FILTER_HSV_MAX);
	createTrackbar("hueUpper", "Controls", FILTER_HSV_MAX);
	createTrackbar("satLower", "Controls", FILTER_HSV_MAX);
	createTrackbar("valLower", "Controls", FILTER_HSV_MAX);
	createTrackbar("filterBackgroundOn", "Controls", 1);
	createTrackbar("filterIterations", "Controls",  FILTER_ITERATIONS_MAX);
	createTrackbar("filterKernel", "Controls", FILTER_KERNEL_MAX); // rect = 0, ellipse = 1, cross = 2
	createTrackbar("filterKernelSize", "Controls", FILTER_KERNEL_SIZE_MAX);
	createTrackbar("filterBlurOn", "Controls", 1);
	createTrackbar("filterSigma", "Controls", FILTER_SIGMA_MAX);

	// Hough detection trackbars
	createTrackbar("houghOn", "Controls", 1);
	createTrackbar("houghThreshUpper", "Controls", HOUGH_THRESH_MAX);
	createTrackbar("houghThreshLower", "Controls", HOUGH_THRESH_MAX);
	createTrackbar("houghMinRadius", "Controls", HOUGH_RADIUS_MAX);
	createTrackbar("houghMaxRadius", "Controls", HOUGH_RADIUS_MAX);

	// Contour detection trackbars
	createTrackbar("contourMaxArea", "Controls", CONTOUR_MAX_AREA_MAX);
	createTrackbar("contourMinArea", "Controls", CONTOUR_MIN_AREA_MAX);
}

void BallTracker::createTrackbar(const char* parameterName, const char* windowName, int maxValue) {
	tempInt = trackingParameters->getParameter(parameterName);
	cv::createTrackbar(parameterName, windowName, &tempInt, maxValue, sliderCallback,
		createSliderCallbackData(trackingParameters, parameterName));
}

void BallTracker::filterFrame(cv::Mat &src, cv::Mat &dst) {
	// Filter with background subtraction
	cv::Mat frameGray(src.size(), src.type());
	cv::cvtColor(src,frameGray, cv::COLOR_BGR2BGRA);
	backgroundSubtractor->apply(frameGray,backgroundMask,0);

	// Filter with HSV thresholds
	cv::Mat frameHsv(src.size(), src.type());
	cv::cvtColor(src,frameHsv, cv::COLOR_BGR2HSV);
	cv::Scalar hsvRangeLower = cv::Scalar(trackingParameters->getParameter("hueLower"),
		trackingParameters->getParameter("satLower"),
		trackingParameters->getParameter("valLower"), 0);
	cv::Scalar hsvRangeUpper = cv::Scalar(trackingParameters->getParameter("hueUpper"),
		trackingParameters->getParameter("satUpper"),
		trackingParameters->getParameter("valUpper"), 0);

	cv::inRange(frameHsv, hsvRangeLower, hsvRangeUpper, dst);

	// Combine with background filtering if enabled
	if (trackingParameters->getParameter("filterBackgroundOn"))
		dst = dst & backgroundMask;

	// Filter image with dilation, guassian blur, and erosion
	int iterations = trackingParameters->getParameter("filterIterations");
	int kernelSize = trackingParameters->getParameter("filterKernelSize");

	
	// Gaussian blur if enabled
	if (trackingParameters->getParameter("filterBlurOn"))
		cv::GaussianBlur(dst, dst, cv::Size(0,0), std::max(trackingParameters->getParameter("filterSigma"),1),
			std::max(trackingParameters->getParameter("filterSigma"),1));
	
	// Dilate and Erode
	cv::Mat kernel = cv::getStructuringElement(trackingParameters->getParameter("filterKernel"), cv::Size(kernelSize,kernelSize));
	for (uint8_t i = 0; i < iterations; i++)
		cv::erode(dst, dst, kernel); 
	for (uint8_t i = 0; i < iterations; i++)
		cv::dilate(dst, dst, kernel); 

}

bool BallTracker::detectBall(cv::Mat &frame, cv::Point2i &center, int &radius) {
	bool found = false;
	radius = 0;
	center.x = 0;
	center.y = 0;

	if (trackingParameters->getParameter("houghOn")) {
		std::vector<cv::Vec3f> circles;
		HoughCircles(frame, circles, cv::HOUGH_GRADIENT, 1,
			trackingParameters->getParameter("houghMinDist"), std::max(trackingParameters->getParameter("houghThreshUpper"),1),
			std::max(trackingParameters->getParameter("houghThreshLower"),1), trackingParameters->getParameter("houghMinRadius"),
			trackingParameters->getParameter("houghMaxRadius"));

		// Find the largest circle
		float maxRadius = 0.0;
		float x = 0;
		float y = 0;

		for (size_t i = 0; i < circles.size(); i++) {
			x = circles[i][0];
			y = circles[i][1];
			if (circles[i][2] > maxRadius) {
				maxRadius = circles[i][2];
				found = true;
				center.x = int(x + 0.5f);
				center.y = int(y + 0.5f);
				radius = int(maxRadius + 0.5f);
			}
		}
	} // hough circle detection
	else {
		// Contour detection
		cv::Mat temp;
		frame.copyTo(temp);
		std::vector<std::vector<cv::Point>> contours;
		std::vector<cv::Vec4i> hierarchy;

		cv::findContours(temp,contours, hierarchy, cv::RETR_CCOMP, cv::CHAIN_APPROX_SIMPLE);

		// Look for largest moment
		double maxArea = 0.0;
		float x = 0.0, y = 0.0;
		int numObjects = hierarchy.size();
		if (numObjects > DEFAULT_CONTOUR_MAX_OBJECTS) {
			if (debug)
				std::cout << "Contour detection found too many objects!" << std::endl;
		}
		else if (numObjects > 0) {
			// Use hierarchy to determine next 
			for (int i = 0; i >= 0; i = hierarchy[i][0]) {
			//for (int i = 0; i < numObjects; i++) {
				cv::Moments moment = cv::moments((cv::Mat)contours[i]);
				double objectArea = moment.m00;
				double objectRadius = sqrt(objectArea/M_PI);
				if (objectArea > trackingParameters->getParameter("contourMinArea")
					&& objectArea < trackingParameters->getParameter("contourMaxArea")
					&& objectArea > maxArea) {
						found = true;
						maxArea = objectArea;
						center.x = (int)(moment.m10/objectArea);
						center.y = (int)(moment.m01/objectArea);
						radius = (int)objectRadius;
				}

			}
		}
	} // contour detection

	return found;
}


void BallTracker::drawBall(cv::Mat &frame, cv::Point2i center, int outerRadius, int innerRadius,
	const cv::Scalar& innerColor, const cv::Scalar& outerColor, const cv::Scalar& crosshairColor) {
	// Draw circles
	cv::circle(frame,center,outerRadius,outerColor,1);

	// Draw crosshairs and inner dot if ball is large enough
	if (outerRadius > 7) { 
		cv::circle(frame,center,innerRadius,innerColor,-1); // draw inner dot
		int crosshairSize = (int)((float)outerRadius*2.2f);
		int crosshairThickness = 1;
		int crosshairOffset = crosshairSize/2;
		cv::line(frame,cv::Point(center.x - crosshairOffset,center.y),
			cv::Point(center.x + crosshairOffset,center.y),
			crosshairColor, crosshairThickness);
		cv::line(frame, cv::Point(center.x, center.y - crosshairOffset),
			cv::Point(center.x, center.y + crosshairOffset),
			crosshairColor, crosshairThickness);
	}
	// Add text next to ball: (x,y, radius px)
	char buf[250];
	double textScale = 0.8f;
	int textThickness = 1;
	cv::Scalar textColor(0,255,0);
	sprintf(buf,"(x: %u, y:%u), %u px",center.x, center.y, outerRadius);
	cv::putText(frame, std::string(buf), cv::Point(center.x + outerRadius, center.y + outerRadius),
		cv::FONT_HERSHEY_PLAIN, textScale, textColor, textThickness);
}

/*----------------------- Non-member Functions ------------------------*/
void sliderCallback(int val, void *userData) {
	SliderCallbackData data = (SliderCallbackData)userData;
	std::string parameterName(data->parameterName);
	data->obj->changeParameter(val, parameterName);
}

SliderCallbackData createSliderCallbackData(TrackingParameters *obj, const char *name) {
	SliderCallbackData data = new _sliderCallbackData();
	data->obj = obj;
	data->parameterName = name;
	return data;
}